core/src/main/java/com/linecorp/armeria/client/logging/AbstractLoggingClient.java

```diff
@@ -115,13 +117,20 @@
         this.responseTrailersSanitizer = requireNonNull(responseTrailersSanitizer, "responseTrailersSanitizer");
         this.responseCauseSanitizer = requireNonNull(responseCauseSanitizer, "responseCauseSanitizer");
         this.sampler = requireNonNull(sampler, "sampler");
+        this.failedSampler = requireNonNull(failedSampler, "failedSampler");
     }
 
     @Override
     public final O execute(ClientRequestContext ctx, I req) throws Exception {
-        if (sampler.isSampled(ctx)) {
-            logWhenComplete(logger, ctx, requestLogger, responseLogger);
-        }
+        ctx.log().whenComplete().thenAccept(log -> {
+            if (log.responseCause() == null) {
```

Comment:
How about also adding a Predicate that determines if a request is failed or not?
Users might want to `failureSampler` when the status is 5XX.

Reply:
`MetricCollecting{Service,Client}` has its own success function.
https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/client/metric/MetricCollectingClient.java#L68
If we also provide another predicate separately, a request could be considered a different result depending on the decorators.
Should we need to support a `Client` level and server level `success function?
```java
Server.builder()
      .successFunction(log -> ...)

WebClient.builder()
      .successFunction(log -> ...)
```

Reply:
I completely missed that. 😅 
Yeah, I think it's a great idea. 👍

Reply:
You mean adding successFunction to ServerBuilder? So `MetricCollectingServiceBuilder`  also need to check if ServerBuilder's successFunction if there is no successFunction at MetricCollectingBuilder?

Reply:
> You mean adding successFunction to ServerBuilder? 

Right.

> check if ServerBuilder's successFunction if there is no successFunction at MetricCollectingBuilder?

I think we can give `ServerBuilder.successFunction()` lower priority.

- Promote the default success function into the default function at `ServerBuilder` and `ClientBuilder` level.
- Deprecate `MetricCollectingClientBuilder.successFunction()`.
- Use `ServiceRequestContext.config().successFunction()` or `ClientRequestContext.options().sucessFunction()`
  if a custom success function is not specified. In short, if a custom success function is specified in 
  `MetricCollecting{Service,Client}`, ignore successFunction() in `RequestContext`.

Reply:
Additionally, it would be nice to configure a success function at `Service` level.

Reply:
> Additionally, it would be nice to configure a success function at Service level.
About Service level is ServiceBindingBuilder? 


Adding this successFunction looks ok for me, but curious if it's appropriate due to Server it self is not using it but decorators/services lookup through  ServerConfig. Or just let user pass the same `BiPredicate` for decorators.

Reply:
> About Service level is ServiceBindingBuilder?

I think we can follow how the `BlockingTaskExecutor` is configured.

> if it's appropriate due to Server it self is not using it but decorators/services lookup through ServerConfig. Or just let user pass the same BiPredicate for decorators.

It should be passed via `ServiceConfig`. Users can set a custom success function per service.

Reply:
> I think we can follow how the BlockingTaskExecutor is configured.
> It should be passed via ServiceConfig. Users can set a custom success function per service.  

I see.

And one question is if it's good idea to add this to `ServiceConfig`? For example if armeria have another built-in decorators with other `Function` or `Predicate`, will we add all of them to `ServiceConfig` or keep it at Decorator level. 

```
Server.builder()
      .successFunction((requestContext, requestLog) -> ...) // server level
      .blockingTaskExecutor(12)  // server level
      .route()
      .successFunction((requestContext, requestLog) -> ...)  // service level
      .blockingTaskExecutor(12)   // service level
      .build(thriftService)
```

or

```

BiPredicate<? super RequestContext, ? super RequestLog> successFunction = (requestContext, requestLog)->{...};

Server.builder()
          .service("/", service.decorate(LoggingService.newDecorator().successFunction(successFunction)
                                         .decorate(MetricCollectingService.builder( MeterIdPrefixFunction.ofDefault(""))
                                                                      .successFunction(successFunction)
                                                                      .newDecorator())));
```

Other user designed decorators can utilize this also, but not sure if it's easy to be aware of general successFunction setting through ServerBuilder/ServiceBindingBuilder.

Reply:
> but not sure if it's easy to be aware of general successFunction setting through ServerBuilder/ServiceBindingBuilder.

Agreed. I think it has pros and cons. It doesn't look intuitive but users can easily configure the success function at the service level which is a bit verbose with decorators.

`ServiceConfig.transientServiceOptions()` is a similar case. It is configured through Server/Service builder and used in some built-in decorators. 
https://github.com/line/armeria/blob/aa647fae9278fa54c4f303c29e1db8f058fbaf78/core/src/main/java/com/linecorp/armeria/server/ServiceConfig.java#L293
https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/server/metric/MetricCollectingService.java#L91